Allow defining an explicit `docker build` context instead of a directory

gdemonet opened the pull request:
**Component**: build

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->

**Context**: See #3271

**Summary**:

- Fix an issue with dependency checking for `LocalImage` targets, and replace this check with a dynamic calculation of these dependencies.
- Introduce an `ExplicitContext` object to prepare a tar archive in memory before sending it as a build context to Docker client.

**Acceptance criteria**: Builds still pass, no flakiness is observed on `_image_build:metalk8s-ui`.


---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #3271

<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

bert-e commented:
# Hello gdemonet,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.8/bugfix/3271-manually-tar-ui-build-context` with contents from `bugfix/3271-manually-tar-ui-build-context`
and `development/2.8`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.8/bugfix/3271-manually-tar-ui-build-context origin/development/2.8
 $ git merge origin/bugfix/3271-manually-tar-ui-build-context
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.8/bugfix/3271-manually-tar-ui-build-context
```

gdemonet committed:
build: Calculate dependencies of LocalImage tasks

The previous approach consisted in:
- a static listing of these deps in the `file_dep` task attribute
- a dynamic verification from parsing the Dockerfile

This dynamic verification was however broken, because it used the
Dockerfile parent directory instead of the build context for listing the
dependencies.

When fixing this issue, we realized that a dynamic check couldn't work
for dependencies generated by other tasks, since the static `file_dep`
could not see them at task creation time.

To work around this limitation, and make the whole experience more
comfortable to developers, instead of checking, we use `doit`'s
`calc_dep` task attribute to dynamically compute dependencies using a
dedicated task, which can then wait for other tasks to complete.

See: #3271
See: #926

gdemonet committed:
build: Use explicit build context for metalk8s-ui

This will prevent the Docker client from tarring up the whole `_build`
directory (and sometimes failing when multiple processes are writing to
this directory).

Fixes: #3271

gdemonet commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.8/bugfix/3271-manually-tar-ui-build-context` with contents from `bugfix/3271-manually-tar-ui-build-context`
and `development/2.8`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.8/bugfix/3271-manually-tar-ui-build-context origin/development/2.8
 $ git merge origin/bugfix/3271-manually-tar-ui-build-context
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.8/bugfix/3271-manually-tar-ui-build-context
```

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.9/bugfix/3271-manually-tar-ui-build-context` with contents from `w/2.8/bugfix/3271-manually-tar-ui-build-context`
and `development/2.9`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.9/bugfix/3271-manually-tar-ui-build-context origin/development/2.9
 $ git merge origin/w/2.8/bugfix/3271-manually-tar-ui-build-context
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.9/bugfix/3271-manually-tar-ui-build-context
```

bert-e commented:
# Conflict




A conflict has been raised during the update of
integration branch `w/2.8/bugfix/3271-manually-tar-ui-build-context` with contents from `bugfix/3271-manually-tar-ui-build-context`
and `development/2.8`.

Please resolve the conflict on **the integration branch** (`w/2.8/bugfix/3271-manually-tar-ui-build-context`).


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout w/2.8/bugfix/3271-manually-tar-ui-build-context
 $ git pull  # or "git reset --hard origin/w/2.8/bugfix/3271-manually-tar-ui-build-context"
 $ git merge origin/development/2.8
 $ # <intense conflict resolution>
 $ git commit
 $ git merge origin/bugfix/3271-manually-tar-ui-build-context
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.8/bugfix/3271-manually-tar-ui-build-context
```

bert-e commented:
# History mismatch



Merge commit #d3206e3468f1d0e43de97c25bd57b9c15e84a477 on the integration branch
`w/2.8/bugfix/3271-manually-tar-ui-build-context` is merging a branch which is neither the current
branch `bugfix/3271-manually-tar-ui-build-context` nor the development branch
`development/2.8`.

It is likely due to a rebase of the branch `bugfix/3271-manually-tar-ui-build-context` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.9/bugfix/3271-manually-tar-ui-build-context` with contents from `w/2.8/bugfix/3271-manually-tar-ui-build-context`
and `development/2.9`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.9/bugfix/3271-manually-tar-ui-build-context origin/development/2.9
 $ git merge origin/w/2.8/bugfix/3271-manually-tar-ui-build-context
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.9/bugfix/3271-manually-tar-ui-build-context
```

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/3271-manually-tar-ui-build-context` into
`development/2.7`
* [w/2.8/bugfix/3271-manually-tar-ui-build-context](https://github.com/scality/metalk8s/compare/development/2.8...w/2.8/bugfix/3271-manually-tar-ui-build-context) will be merged into `development/2.8` 
* [w/2.9/bugfix/3271-manually-tar-ui-build-context](https://github.com/scality/metalk8s/compare/development/2.9...w/2.9/bugfix/3271-manually-tar-ui-build-context) will be merged into `development/2.9` 
* [w/2.10/bugfix/3271-manually-tar-ui-build-context](https://github.com/scality/metalk8s/compare/development/2.10...w/2.10/bugfix/3271-manually-tar-ui-build-context) will be merged into `development/2.10` 



The following branches will **NOT** be impacted:

* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nicolast

* @chengyanjin

* @alexandre-allard-scality

* @thomasdanan

* @gdemonet

* @teddyandrieux

* @lucieleonard

* @jbwatenbergscality

gdemonet commented:
/approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nicolast

* @chengyanjin

* @alexandre-allard-scality

* @thomasdanan

* @gdemonet

* @teddyandrieux

* @lucieleonard

* @jbwatenbergscality










*The following options are set:* **approve**

alexandre-allard commented on buildchain/buildchain/targets/local_image.py:
```diff
@@ -212,3 +159,19 @@ def _do_save(self) -> List[types.Action]:
+        if dep.is_dir():
+            return list(coreutils.ls_files_rec(dep))
+        # Globs - `*` or specific e.g. `*.py`, `*.repo`
+        return [
+            sub_path
+            for sub_path in dep.parent.glob(dep.name)
+            if sub_path.is_file()
+        ]
```
What happens if I have the following:
```
COPY src/* dest/
```
And in `src`, I have multiple level of directories, e.g.:
```
src
├── foo
│   └── bar
│       └── baz
└── banana
```
We don't add `src/foo/bar/baz` as a dependency, no ?

Note: there is no such case in any of our Dockerfile
Note2: it was already the case before this PR

gdemonet replied:
I didn't change the behaviour for parsing such glob patterns, so I don't really know, and didn't test. I would assume `pathlib.Path.glob` works roughly the same as Docker does when parsing those statements. If it doesn't, I'd suggest we open a tracking issue and tackle it in another PR :wink:

gdemonet replied:
Just checked, and indeed, there is a difference between the two. I guess this can be easily improved, by doing something like:

```python
deps = []
for sub_path in dep.parent.glob(dep.name):
    if sub_path.is_file():
        deps.append(sub_path)
    elif sub_path.is_dir():
        deps.extend(coreutils.ls_files_rec(sub_path))
    else:
        # do something? or not?
        continue
return deps

gdemonet replied:
See #3359

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.7`

* :heavy_check_mark: `development/2.8`

* :heavy_check_mark: `development/2.9`

* :heavy_check_mark: `development/2.10`



The following branches will **NOT** be impacted:

* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.7`

* :heavy_check_mark: `development/2.8`

* :heavy_check_mark: `development/2.9`

* :heavy_check_mark: `development/2.10`



The following branches have **NOT** changed:

* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`



Please check the status of the associated issue None.

Goodbye gdemonet.
